floor/floor/controller/controller.py

```diff
@@ -149,55 +126,46 @@ def run_forever(self):
     def run_one_frame(self):
         if not self.playlist.is_running():
             # If the playlist is stopped/paused, sleep a bit then restart the loop
-            self.sleeper(0.5)
+            self.clocksource.sleep(0.5)
             return
 
         self.init_loop()
-        self.check_playlist()
+        self.prepare()
         self.generate_frame()
         self.transfer_data()
         self.delay()
 
     def init_loop(self):
         self.frame_start = self.clocksource.time()
 
-    def check_playlist(self):
-        item = self.playlist.get_current()
-        if not item:
-            return
-
-        processor_name, args = item['name'], item['args']
-        if processor_name not in self.processors:
-            logger.error('Unknown processor "{}"; removing it.'.format(processor_name))
-            self.playlist.remove(self.playlist.position)
-            return
-
-        if processor_name and (processor_name, args) != (self.current_processor, self.current_args):
-            logger.debug('Loading processor {}'.format(processor_name))
-            self.set_processor(processor_name, args)
-            # Make sure the processor is limited to the bit depth of the driver
-            self.processor.set_max_value(self.max_effective_led_value)
+    def prepare(self):
+        for layer in self._iter_enabled_layers():
+            layer.prepare()
 
     def generate_frame(self):
-        if not self.processor:
-            return
-
         context = RenderContext(
             clock=self.frame_start,
             downbeat=self.downbeat,
             weights=self.get_weights(),
             bpm=self.bpm,
         )
 
-        try:
-            leds = self.processor.get_next_frame(context)
-        except KeyboardInterrupt:
-            raise
-        except Exception:
-            logger.exception('Error generating frame for processor {}'.format(self.processor))
-            logger.warning('Removing processor due to error.')
-            self.playlist.remove(self.playlist.position)
-        else:
+        leds = None
+        last_leds = None
+        for layer in self._iter_enabled_layers():
+            leds = layer.render(context, leds=last_leds)
+            if leds and last_leds:
+                # Copy the returned buffer, since we'll likely be blending into it
+                # and cannot assume ownership.
+                leds = leds[:]
+                for idx, last_pixel in enumerate(last_leds):
+                    leds[idx] = blend_pixel_nonalpha(last_pixel, leds[idx])
+            last_leds = leds
+
+        # If no layers are enabled, `leds` will be None and we shouldn't update the driver.
+        # TODO(mikey): Should we render a 'default' pattern in this case?
```

Comment:
I don't think so;  it should be possible for the floor to just be "off" and not displaying anything ... unless I'm misunderstanding this.